llama.cpp

```diff
@@ -5625,6 +5729,96 @@ struct llm_build_context {
         return gf;
     }
 
+    struct ggml_cgraph * build_bert() {
+        struct ggml_cgraph * gf = ggml_new_graph_custom(ctx0, LLAMA_MAX_NODES, false);
+
+        const int64_t n_embd_head = hparams.n_embd_head_v;
+        GGML_ASSERT(n_embd_head == hparams.n_embd_head_k);
+        GGML_ASSERT(n_embd_head == hparams.n_rot);
+
+        struct ggml_tensor * cur;
+        struct ggml_tensor * inpL;
+
+        // get input vectors with right size
+        struct ggml_tensor * inp_type = ggml_view_1d(ctx0, lctx.inp_type, n_tokens, 0);
+        struct ggml_tensor * inp_pos = ggml_view_1d(ctx0, lctx.inp_pos, n_tokens, 0);
+        struct ggml_tensor * inp_sum = ggml_view_1d(ctx0, lctx.inp_sum, n_tokens, 0);
+
+        // construct input embeddings (token, type, position)
+        inpL = llm_build_inp_embd(ctx0, hparams, batch, model.tok_embd, lctx.inp_tokens, lctx.inp_embd, cb);
+        inpL = ggml_add(ctx0, ggml_get_rows(ctx0, model.type_embd, inp_type), inpL);
```

Comment:
Since `inp_type` is always zero, is there a reason for using a `get_rows` here instead of `model.type_embd` directly (or a view of the first row)? It seems that this `get_rows` is being used only to broadcast `type_embd`.

Reply:
I've been trying to implement your suggestion, but unfortunately I have been unable to duplicate the behavior of ggml_get_rows which both repeats and converts to F32.
- The only broadcastable add op is F32/F32, so I must repeat or convert.
- There is no F32/F16->F32 add op, only F16/F32->F16, so I cannot simply swap the arguments and repeat src1.
- I don't know how to convert to F32 without a ggml_cpy to a new tensor, which AFAIU hasn't been allowed since #5100.

Reply:
Yup, hard-coded a view to `model.type_embd` row zero.

Reply:
For this to work, these weights would need to be exported as F32 like all other biases.